addons/account/models/product.py

```diff
@@ -89,13 +90,17 @@ def _get_tax_included_unit_price(self, company, currency, document_date, documen
                 product_taxes = product.taxes_id.filtered(lambda x: x.company_id == company)
             elif document_type == 'purchase':
                 product_taxes = product.supplier_taxes_id.filtered(lambda x: x.company_id == company)
+
         # Apply unit of measure.
         if product_uom and product.uom_id != product_uom:
             product_price_unit = product.uom_id._compute_price(product_price_unit, product_uom)
 
         # Apply fiscal position.
-        if product_taxes and fiscal_position:
-            product_taxes_after_fp = fiscal_position.map_tax(product_taxes)
+        if (product_taxes or line_taxes) and fiscal_position:
```

Comment:
You should use `is None` everytime. This is hard to understand.. What about
```py
if line_taxes is not None:
    taxes_after_fp = line_taxes
elif product_taxes is not None:
    taxes_after_fp = fiscal_position.map_tax(product_taxes)
else:
    taxes_after_fp = None

if taxes_after_fp is not None and set(product_taxes.ids) != set(product_taxes_after_fp.ids):
    ....
```
Note: Are you sure the last condition is still true? Should we apply the fiscal position on the price_unit only when `product_taxes is not None` ? This need some comments because I don't understand clearly what you are trying to fix...

Reply:
Once taxes are computed on the SOL (or POL as well I suppose), we shouldn't base the computation on the product taxes anymore and only consider the taxes on the SOL (including taxes added by hand potentially)